src/utils/parser/getScopes.js

```diff
@@ -0,0 +1,280 @@
+}
+
+function _toGeneratedScopes(
+  children: TempScope[],
+  sourceId: SourceId
+): ?(GeneratedScope[]) {
+  if (!children || children.length === 0) {
+    return undefined;
+  }
+  return children.map(scope => {
+    let bindings = Object.create(null);
+    Object.keys(scope.names).forEach(n => {
+      switch (scope.names[n].kind) {
+        case "var":
+        case "let":
+        case "const":
+        case "param":
+          bindings[n] = scope.names[n].refs.map(location => {
+            return _fromBabelLocation(location, sourceId);
+          });
+          break;
+      }
+    });
+    return {
+      start: _fromBabelLocation(scope.loc.start, sourceId),
+      end: _fromBabelLocation(scope.loc.end, sourceId),
+      type: scope.type.toLowerCase(),
+      bindings,
+      children: _toGeneratedScopes(scope.scopes, sourceId)
+    };
+  });
+}
+
+function _parseJSScopes(source: Source) {
+  let parent: TempScope = _createTempScope("Global", null, null);
+  let savedParents: WeakMap<NodePath, TempScope> = new WeakMap();
+  traverseAst(source, {
+    enter(path: NodePath) {
+      const tree = path.node;
+      const location = path.node.loc;
+      if (path.isProgram()) {
+        savedParents.set(path, parent);
+        const scope = _createTempScope("Script", parent, location);
+        parent = scope;
+        return;
+      }
+      if (
+        path.isFunctionDeclaration() ||
+        path.isFunctionExpression() ||
+        path.isArrowFunctionExpression()
+      ) {
+        savedParents.set(path, parent);
+        const scope = _createTempScope("Function", parent, location);
+        if (isNode(tree.id, "Identifier")) {
+          const functionName = { kind: "fn", refs: [] };
+          _getFunctionScope(parent).names[tree.id.name] = functionName;
+          scope.names[tree.id.name] = functionName;
+        }
+        tree.params.forEach(param => {
+          _parseDeclarator(param, scope, "param");
+        });
+        parent = scope;
+        return;
+      }
+      if (path.isForInStatement() || path.isForStatement()) {
+        const init = tree.init || tree.left;
+        let scope = parent;
+        if (
+          isNode(init, "VariableDeclaration") &&
+          (init.kind === "let" || init.kind === "const")
+        ) {
+          savedParents.set(path, parent);
+          scope = _createTempScope("For", parent, location);
+          parent = scope;
+        }
+        return;
+      }
+      if (path.isCatchClause()) {
+        savedParents.set(path, parent);
+        const scope = _createTempScope("Catch", parent, location);
+        _parseDeclarator(tree.param, scope, "param");
+        parent = scope;
+        return;
+      }
+      if (path.isBlockStatement()) {
+        const hasLet = tree.body.some(node => {
+          if (!isNode(node, "VariableDeclaration")) {
+            return false;
+          }
+          return node.kind === "let" || node.kind === "const";
+        });
+        let scope = parent;
```

Comment:
the comments written here: https://github.com/devtools-html/debugger.html/pull/3852/files/e0f9e89e7ee904fbd1d1a17a0d801e8f7a1965dc#diff-dcf4c6f392912b09dabbbf7f6d2436b4

can also apply here